### DISCUSSION TITLE
is there full Hotkey and Hotstring support?

### DISCUSSION THREAD

**Discussion by levavft** [spyoungtech/ahk#155] [Q&A] [answered]
Hi, I was wondering If there was support for everything that has to do with hotkeys and hotstrings?
for example, support for hotstring options (:?o:somestring::someotherstring)

**Comment by spyoungtech** [comment, marked as answer]
Almost. Technically, every functionality of AHK is supported through the `.run_script` API. Hotkeys are supported directly in the API (`ahk.hotkey`), but Hotstrings for text replacement (e.g. `::btw::by the way
`) are not _directly_ supported by the API. 

Hotstrings that run scripts can work by using the hotkey API.

I'll make a note to implement hotstrings for simple text replacement in the near future.

Using the Hotkey API, you can do this:

```python
import time
from ahk import AHK
ahk = AHK()
hotstring = ahk.hotkey('::btw', 'MsgBox you typed "btw".')
hotstring.start()
print('hotstring active! Press ctrl+c to stop')

# add a loop to stop the Python script from exiting
while True:
    try:
        time.sleep(0.5)
    except KeyboardInterrupt:
        hotstring.stop()  # stop the hotstring script
        break
```

This is equivalent to the AHK script:

```ahk
::btw::
MsgBox You typed "btw".
return
```

You can also implement hotstrings yourself using `run_script` like:

```python
ahk = AHK()
hotstring = ahk.run_script('::btw::by the way', blocking=False)
print('hotstring active! Press ctrl+c to stop')
# add a loop to stop the Python script from exiting
while True:
    try:
        time.sleep(0.5)
    except KeyboardInterrupt:
        hotstring.terminate()  # stop the hotstring script
        break
```

**Comment by levavft** [comment]
thank you very much! and there is no issue with running multiple hotstrings in a single script, or, multiple scripts of a single hotstring right?

e.g. 
hotstring = ahk.run_script('::btw::by the way\n::lol::laugh out loud', blocking=False)
hs2 = ahk.run_script('::M.s.c::Master of science', blocking=False)

should work right?

**Comment by spyoungtech** [comment]
There's no particular issue with this, no. Though something to be aware of: every time you call `run_script` you're creating a new process. So, in general, it would be better to shove them all into a single script if you have dozens of them.

**Comment by levavft** [comment]
ty so much!
